design_drafts/controller_execution_management.md

```diff
@@ -0,0 +1,125 @@
+A typical use case for this is a safety `JointLimitsController` which is being linked after a `JointEffortController` to enforce an operation within appropriate min/max torque values.
+Another use case could be to clamp a `DiffDriveController` velocity values to optimize for battery lifetime or top acceleration.
+
+In order to realize such a controller chaining, the output of one's controller must be linked as the input to its consecutive controller.
+This further motivates a relaxation that resources such as the `JointHandle` can only be uniquely claimed.
+
+### Sequential Controller Container
+
+Instead of loading each controller individually as an independent controller, we propose a parent container which holds and connects controllers which are configured in a specific order.
+The container has a single input and a single output which is being scoped within the container.
+That is, all controllers loaded within a container don't see any input/output states outside the container.
+The input and output states can then be safely re-routed and linked between individual controllers.
+After all controllers are executed, the final output state is being returned from the container.
+
+The image below shows how this execution model could look like:
+![sequential_execution](images/sequential_execution.png "Sequential Execution Container")
+
+## Parallel Execution of Controllers
+
+Opposite to the sequential order, there are use cases where it makes sense to operate controllers in a parallel fashion.
+`Controller_1`, `Controller_2`, …, `Controller_3` are being executed in parallel/isolated and their output state is being fused with a specific function.
+
+### Separation of Tasks
+
+Having controllers being run in parallel allows a better separation of tasks per controller.
+One easy to grasp example (also if not the most sensible) is a simple `PID-Controller`.
+We could break down each gain (`P`, `I`, `D`) in its own controller and let them calculate in isolation.
+By being able to execute multiple controllers in parallel, the calculation of the actual output can then be to simply add the `P`, `I`, `D` outputs - according to the control law of `p * (x_d - x) + p * (x_d - x) / dt  + i * sum(x_d -x) * dt`.
+A second use case could be again a differential drive controller which computes rotational and translational velocities and a second controller calculates dynamic friction values which then again can be fused together.
+
+### Parallel Controller Container
+
+Just like in the sequential execution, we propose having a container structure which holds a set of loaded controllers which are being executed in parallel and primarily independent from each other.
+When all controllers were executed, each individual calculated values are being joined.
+As already mentioned previously, this parallel container can be initialized with a custom designed operation which dictates how all commonly calculated values (e.g. `torques`) are being combined.
```

Comment:
```suggestion
As already mentioned previously, each controller group type can be initialized with a custom strategy which dictates how all commonly calculated values (e.g. `torques`) are being combined.
```

Reply:
This strategy is not applicable for a sequential controller group, but exclusively for the parallel one.